flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/connector/source/LookupRuntimeProviderContext.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.connector.source;
+
+import org.apache.flink.table.connector.source.DynamicTableSource.DataStructureConverter;
+import org.apache.flink.table.connector.source.LookupTableSource;
+import org.apache.flink.table.data.util.DataFormatConverters.DataFormatConverter;
+import org.apache.flink.table.types.DataType;
+
+import static org.apache.flink.table.data.util.DataFormatConverters.getConverterForDataType;
+
+/**
+ * Implementation of {@link LookupTableSource.Context}. Currently we delegate
+ * {@link #createDataStructureConverter} to {@link DataFormatConverter}.
+ *
+ * <p>In the future, we can code generate the implementation of {@link #createDataStructureConverter}
+ * for better performance.
+ */
+public class LookupRuntimeProviderContext implements LookupTableSource.Context {
+
+	private final int[][] lookupKeys;
+
+	public LookupRuntimeProviderContext(int[][] lookupKeys) {
+		this.lookupKeys = lookupKeys;
```

Comment:
Why is `lookupKeys` two-dimensional array.

Reply:
See the Javadoc `org.apache.flink.table.connector.source.LookupTableSource.Context#getKeys`, it is in order to support nested lookup keys.

Reply:
But IMO, two-dimensional array is not enough, what if there is a column `a.b.c` ?

Reply:
It can be expressed. If we have a lookup table , and the lookup key is `r.f.q2`
```sql
CREATE TABLE t1 (r ROW< d DOUBLE, f ROW<q1 STRING, q2 BOOLEAN> >);
```
the array will be `[[0, 1, 1]]` which is equivalent to `[["r", "f", "q2"]]`.

Reply:
The array is a path for selection.